modules/nf-commons/src/main/nextflow/util/HashBuilder.java

```diff
@@ -344,12 +353,25 @@ public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) throws IO
                 public FileVisitResult preVisitDirectory(Path path, BasicFileAttributes attrs) {
                     log.trace("Hash sha-256 dir content [DIR] path={} - base={}", path, base);
                     // the file relative base
-                    if( base!=null )
-                        hasher.putUnencodedChars(base.relativize(path).toString());
-                    hasher.putUnencodedChars(base.relativize(path).toString());
+                    String entry = base != null ? base.relativize(path).toString() : null;
+                    // ↓ wouldn't this cause a null pointer exception when base is null?
+                    String value = base.relativize(path).toString();
+                    
+                    pathShaPairs.add(Map.entry(entry, value));
                     return FileVisitResult.CONTINUE;
                 }
             });
+
+            // sort the file hashes
+            Collections.sort(pathShaPairs, Comparator.comparing(Map.Entry::getKey));
```

Comment:
It would be preferable using `TreeMap` to avoid sorting the entries separately

Reply:
Good point. I swapped the ArrayList for a TreeMap!